docs/HANDLERS.md

```diff
@@ -0,0 +1,376 @@
+# Handlers
+
+In addition to the Applications the toolkit provides a variety of the utilities to perform most common actions of the ship. Following list of handlers includes a prebuild Express routers which performs most common operations of connectors. 
+
+### actionHandler() `incoming`
+This is the simplest requests handler to expose custom logic through custom API POST endpoint. The possible usage is triggering a custom operation (like fetching historical data) or a webhook. Both cases handle data `incoming` into Hull platform. However in case of busy webhook it's better to use `batcherHandler` which automatically group the incoming requests into batches.
+
+```js
+const app = new Hull.App({}).server();
+import { actionHandler } from "hull/lib/utils";
+
+app.use("/fetch-all", actionHandler((ctx, { query, body }) => {
+  const { hull, ship } = ctx;
+
+  const { api_token } = ship.private_settings;
+  const serviceClient = new ServiceClient(api_token);
+  return serviceClient.getHistoricalData()
+    .then(users => {
+      users.map(u => {
+        hull.as({ email: u.email }).traits({
+          new_trait: u.custom_value
+        });
+      });
+    });
+}));
+```
+
+### batcherHandler() `incoming`
+The second `incoming` handler which works in a simillar way as `actionHandler` but it also groups incoming requests into batches of selected size:
+
+```js
+const app = new Hull.App({}).server();
+import { batcherHandler } from "hull/lib/utils";
+
+app.use("/fetch-all", batcherHandler((ctx, requests) => {
+  requests.map(request => {
+    console.log(request); // { query, body }
+  })
+}, {
+  maxSize: 100, // maximum size of the batch
+  maxTime: 1000 // time time in milliseconds to flush batch after the first item came in
+}));
+```
+
+### notifHandler() `outgoing`
+
+NotifHandler is a packaged solution to receive User and Segment Notifications from Hull. It's built to be used as an express route. Hull will receive notifications if your ship's `manifest.json` exposes a `subscriptions` key:
+
+```json
+{
+  "subscriptions" : [ { "url" : "/notify" } ]
+}
+```
+
+Here's how to use it.
+
+```js
+const app = new Hull.App({}).server();
+import { notifHandler } from "hull/lib/utils";
+
+const handler = NotifHandler({
+  userHandlerOptions: {
+    groupTraits: true, // groups traits as in below examples
+    maxSize: 6,
+    maxTime: 10000
+  },
+  onSubscribe() {} // called when a new subscription is installed
+  handlers: {
+    "ship:update": function(ctx, messages = []) {},
+    "segment:update": function(ctx, messages = []) {},
+    "segment:delete": function(ctx, messages = []) {},
+    "user:update" : function(ctx, messages = []) {
```

Comment:
Then you need to rewrite the api. Because "user:update" is singular: It should not return multiple users.

Micro batching is library level plumbing. I don't want to think about this as a developer.
This is less clear and more complex to apprehend than individual messages.

At the library level yes, but don't make me think about them.